pyggi/line/line.py

```diff
@@ -9,20 +9,6 @@ class LineProgram(AbstractProgram):
     def get_engine(cls, file_name):
```

Comment:
Both LineProgram and TreeProgram have only a single method, which is rather pointless.
I think this can now be moved directly in the main program entry point (e.g., improve.py).

Reply:
`get_engine` defines the semantic of each class `LineProgram` and `TreeProgram`. I think these class-engine mappings should be defined inside of PyGGi, not the user-side program.

Reply:
In any case the user still has to chose which engine he wants to use.
Providing the classes `LineProgram` and `TreeProgram` solely containing the `get_engine` method is simply akin is replace an explicit decision (which engine to use) by another, blurrier, explicit decision (which type of program to use).

My last argument is that it is much cleaner to conditionally change a method content that the inheritance of a class, for example for an entry point allowing the user to use a CLI options to use either line or tree granularity level (albeit I see that Python allows some heretical constructs such as `class MyProgram(LineProgram if args.line else TreeProgram)`).

Reply:
Furthermore, I am thinking that removing or renaming tags in the XML tree in order to have a shorter list of modification points will necessarily require sub-classing the XmlEngine to overwrite its `get_content` method (or monkeypatch it, wich is not desirable).
In this case, it is utterly pointless to use TreeProgram because its `get_engine` method will be ignored (to use say MyXmlEngine rather than XmlEngine).

Reply:
I don't think it is desirable to define another Engine inheriting `SrcMLEngine`. We can rather create another (lower) level of subclasses (XmlNode(Tag)Class or something) and connect them in the main engine class as it would be more clear that all the SrcML handling parts are executed via the `SrcMLEngine`. 
Then, we just implement the code manipulators (java, c++, ,.. tree with `SrcMLEngine` and python code with `AstorEngine`) and tie them together to ultimately provide the `TreeProgram` and then users don't even need to care about what's behind `TreeProgram` or what `SrcML` or `Astor` is. So what users can/need to see are only `TreeProgram`(that supports `python`, `java`, and etc.) and the available edit operators.

Reply:
Or, do you mean providing two versions of SrcmlEngine?

Reply:
An engine is a collection of methods describing how to deal with a type of file (and I still think everything should be class methods). In particular, it hides the method that parses the original code and output its representation.
For xml, the representation is a tree with (far too) many tags, and I suppose it will be absolutely necessary to remove or rename some of the tags.
One solution would be to use class variables to setup the class so that this is done automatically during parsing; this is _horrible_.
Another much simpler and cleaner solution would be to subclass the xml engine to provide a new parse method (basically, a `super` and then the modification of the tree), and provide this engine to PyGGI instead.

Reply:
> An engine is a collection of methods describing how to deal with a type of file (**and I still think everything should be class methods**). In particular, it hides the method that parses the original code and output its representation.

Ok, I'll change it.

And why don't we provide only the XML engine with the new parse method? I mean, we can just convert the heavy XML representations into the simpler version of representation in that engine without requiring any subclasses. Or maybe you can just make commits, and then we'll see.